# MATH 210 Introduction to Mathematical Computing

**January 12, 2024**

* Numeric Types: integer, float, complex
* Arithmetic Operations
* Examples: Roots and Taylor Series

## Numeric Types and Arithmetic Operations

There are 3 main numeric types: integers, floats, complex numbers.

The usual arithmetic operations: `+`, `-`, `*`, `/`, `**`.

Integers are whole numbers such as 1 and -5. Add integers:

In [1]:
2005 + 19

2024

In [2]:
2024 - 1998

26

In [3]:
1*2*3*4*5

120

In [4]:
2**2

4

In [5]:
type(42)

int

In [6]:
20/5

4.0

In [7]:
type(4.0)

float

A floating point number (aka float) is a real number in decimal form. Note that division of integers always returns a float.

In [8]:
12/3

4.0

In [9]:
12/5

2.4

In [10]:
2**(0.5)

1.4142135623730951

In [11]:
0.5**0.5

0.7071067811865476

In [12]:
0.1 + 0.2

0.30000000000000004

Use scientific notation for floats.

In [13]:
1e-3

0.001

In [14]:
type(0.01)

float

In [15]:
3.1415926

3.1415926

The complex number $\sqrt{-1}$ is denoted by `1j` in Python.

In [16]:
1+1j

(1+1j)

In [17]:
type(1+1j)

complex

In [18]:
(-1)**0.5

(6.123233995736766e-17+1j)

What is this?! The number `6.123233995736766e-17` is $6.123233995736766 \times 10^{-17} \approx 0$ and so we interpret it as 0. Therefore `(-1)**0.5` is `1j`.

If `x` is an integer and `y` is a float then what is the type of `x*y`?

In [19]:
2*3.14159

6.28318

In [20]:
type(2*3.14159)

float

If `x` and `y` are integers then what is the type of `x**y`?

In [21]:
2**5

32

In [22]:
(-2)**3

-8

In [23]:
3**(-1)

0.3333333333333333

If `y` positive or 0 then `x**y` is an integer. If `y` negative then `x**y` is a float. 

## Examples

### Quadratic Formula

The roots of $ax^2 + bx + c$ are given by the quadratic formula:

$$
\frac{-b \pm \sqrt{b^2 - 4ac}}{2a}
$$

For example, the roots of $x^2 - x - 1$ are:

In [24]:
(1 + ((-1)**2 - 4*1*(-1))**0.5)/(2*1)

1.618033988749895

In [25]:
(1 - ((-1)**2 - 4*1*(-1))**0.5)/(2*1)

-0.6180339887498949

Note that we use parentheses to group arithmetic operations together. We can't use `[ ]` or `{ }`.

In [26]:
[1 + 5**0.5]/2

TypeError: unsupported operand type(s) for /: 'list' and 'int'

In [27]:
{1 + 5}/2

TypeError: unsupported operand type(s) for /: 'set' and 'int'

### Taylor Series

The Taylor series of $f(x)$ centered at $x=a$ is

$$
f(x) = \sum_{k=0}^{\infty} \frac{f^{(k)}(a)}{k!} (x - a)^k
$$

Taylor series of $e^x$ centered at $a=0$ is:

$$
e^x = \sum_{k=0}^{\infty} \frac{1}{k!} (x - 0)^k = \sum_{k=0}^{\infty} \frac{x^k}{k!}
$$

Approximate $e$ by $x=1$ and summing up to $N=5$:

$$
e \approx \sum_{k=0}^{N} \frac{1}{k!}
$$

In [28]:
1 + 1/1 + 1/(1*2) + 1/(1*2*3) + 1/(1*2*3*4) + 1/(1*2*3*4*5)

2.7166666666666663

Approximate $1/e$ by $x=-1$ and summing up to $N=5$:

$$
e^{-1} \approx \sum_{k=0}^{N} \frac{(-1)^k}{k!}
$$

In [29]:
1 - 1/1 + 1/(1*2) - 1/(1*2*3) + 1/(1*2*3*4) - 1/(1*2*3*4*5)

0.3666666666666667

## Variables

Use variable `x` so that we can change the value without editing the whole formula.

In [30]:
x = 1
x**0 + x**1/1 + x**2/(1*2) + x**3/(1*2*3) + x**4/(1*2*3*4) + x**5/(1*2*3*4*5)

2.7166666666666663